File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] lmax = new int[n];
        int[] rmax = new int[n];

        // fill lmax and rmax
        for(int i=1; i<n; i++) {
            lmax[i] = Math.max(height[i-1], lmax[i-1]);
        }
        for(int i=n-2; i>=0; i--) {
            rmax[i] = Math.max(height[i+1], rmax[i+1]);
        }

        int ans = 0;
        for(int i=0; i<n; i++) {
            int min_height = Math.min(lmax[i], rmax[i]);
            if(min_height > height[i]) {
                ans += (min_height - height[i]);
            }
        }
        return ans;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int len = height.size();
        // base case
        if(len <= 2) {
            return 0;
        }

        int *leftMaxTillNow = new int[len];
        int *rightMaxTillNow = new int[len];
        int lMax = 0;
        int rMax = 0;
        // update left array
        for(int i=0; i<len; i++) {
            if(height[i] > lMax) {
                lMax = height[i];
            }
            leftMaxTillNow[i] = lMax;
        }
        // update right array
        for(int i=len-1; i>=0; i--) {
            if(height[i] > rMax) {
                rMax = height[i];
            }
            rightMaxTillNow[i] = rMax;
        }   

        int totalWater = 0;
        for(int i=0; i<len; i++) {
            int minHeight = min(leftMaxTillNow[i], rightMaxTillNow[i]);
            int curr_water = minHeight - height[i];
            totalWater += curr_water;
        }

        return totalWater;
    }
};